src/global/install.rs

```diff
@@ -444,63 +503,122 @@ pub(crate) async fn sync(
+///
+/// This function verifies that all the given specifications are present in the
+/// local environment's prefix records and that there are no extra entries in
+/// the prefix records that do not match any of the specifications.
+fn specs_match_local_environment(
+    specs: &IndexMap<PackageName, MatchSpec>,
+    prefix_records: Vec<PrefixRecord>,
+    platform: Option<Platform>,
+) -> bool {
+    // Check whether all specs in the manifest are present in the installed environment
+    let specs_in_manifest_are_present = specs.iter().all(|(name, spec)| {
+        let Some(prefix_record) = prefix_records
+            .iter()
+            .find(|record| record.repodata_record.package_record.name == *name)
+        else {
+            return false;
+        };
+
+        spec.matches(&prefix_record.repodata_record)
+    });
+
+    if !specs_in_manifest_are_present {
+        return false;
+    }
+
+    // Check whether all packages in the installed environment have the correct package
+    let platform_specs_match_env = prefix_records.iter().all(|record| {
+        let Ok(package_platform) =
+            Platform::from_str(&record.repodata_record.package_record.subdir)
+        else {
+            return true;
+        };
+
+        match package_platform {
+            Platform::NoArch => true,
+            p if Some(p) == platform => true,
+            _ => false,
+        }
+    });
+
+    if !platform_specs_match_env {
+        return false;
+    }
+
+    fn prune_dependencies(
+        remaining_prefix_records: Vec<PrefixRecord>,
+        matched_record: &PrefixRecord,
+    ) -> Vec<PrefixRecord> {
+        let dependencies = &matched_record.repodata_record.package_record.depends;
+        dependencies
+            .iter()
+            .fold(remaining_prefix_records, |mut acc, dependency| {
+                let Some(dependency_name) =
+                    MatchSpec::from_str(dependency, ParseStrictness::Lenient)
+                        .ok()
+                        .and_then(|d| d.name)
+                else {
+                    return acc;
+                };
+
+                let Some(index) = acc.iter().position(|record| {
+                    record.repodata_record.package_record.name == dependency_name
+                }) else {
+                    return acc;
+                };
+
+                let matched_record = acc.remove(index);
+                prune_dependencies(acc, &matched_record)
+            })
+    }
+
+    // Process each spec and remove matched entries and their dependencies
+    let remaining_prefix_records = specs.iter().fold(prefix_records, |mut acc, (name, spec)| {
+        let Some(index) = acc.iter().position(|record| {
+            record.repodata_record.package_record.name == *name
+                && spec.matches(&record.repodata_record)
+        }) else {
+            return acc;
+        };
+        let matched_record = acc.remove(index);
+        prune_dependencies(acc, &matched_record)
+    });
```

Comment:
This algorithm is recursive which might cause a stack overflow if the dependency chain is deep. Try to use a non-recursive algorithm using a work queue instead.

Reply:
Good point, done!

Reply:
Done